File: tests/fixtures/sample-python-project-positive/services/api_gateway/controllers/user_controller.py
"""User controller for handling user-related API requests."""
from flask import request, jsonify
from services.api_gateway.services.user_service import UserService


class UserController:
    """Handles user CRUD operations via the API layer."""

    def __init__(self) -> None:
        self.user_service = UserService()

    def get_all(self) -> tuple:
        """Return all users as JSON."""
        users = self.user_service.find_all()
        return jsonify(users)

    def get_by_id(self, user_id: str) -> tuple:
        """Return a single user by ID."""
        user = self.user_service.find_by_id(user_id)
        if not user:
            return jsonify({"error": "User not found"}), 404
        return jsonify(user)

    def create(self) -> tuple:
        """Create a new user from request data."""
        raw = request.get_json()
        name = str(raw.get("name") or "unknown")
        email = str(raw.get("email") or "unknown")
        validated = {"name": name, "email": email}
        user = self.user_service.create(validated)
        return jsonify(user), 201

    def delete(self, user_id: str) -> tuple:
        """Delete a user by ID."""
        self.user_service.delete(user_id)
        return "", 204

File: tests/fixtures/sample-csharp-project-negative/services/ApiGateway/Controllers/UserController.cs
using ApiGateway.Services;

namespace ApiGateway.Controllers;

public class UserController
{
    private readonly UserService _userService;

    public UserController()
    {
        _userService = new UserService();
    }

    public IResult GetAll()
    {
        var users = _userService.FindAll();
        return Results.Ok(users);
    }

    public IResult GetById(string userId)
    {
        var user = _userService.FindById(userId);
        if (user == null)
        {
            return Results.NotFound(new { error = "User not found" });
        }
        return Results.Ok(user);
    }

    public IResult Create(Dictionary<string, object> data)
    {
        var user = _userService.Create(data);
        return Results.Created($"/users/{user["id"]}", user);
    }

    public IResult Delete(string userId)
    {
        _userService.Delete(userId);
        return Results.NoContent();
    }
}
